File: jar/src/main/java/jline/api/trace/Mtrace_pc.java
package jline.api.trace;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.TreeSet;

import jline.util.matrix.Matrix;

public final class Mtrace_pc {
    private Mtrace_pc() {}

    /**
     * Computes the probabilities of arrival for each class.
     *
     * @param T the inter-arrival times (ignored, for orthogonality with other APIs)
     * @param C the class labels
     * @return column vector of probabilities for each class
     */
    public static Matrix mtrace_pc(double[] T, int[] C) {
        TreeSet<Integer> distinct = new TreeSet<Integer>();
        for (int c : C) {
            distinct.add(c);
        }
        List<Integer> labels = new ArrayList<Integer>(distinct);
        Collections.sort(labels);
        int m = labels.size();

        double[] pc = new double[m];

        for (int i = 0; i < m; i++) {
            int label = labels.get(i);
            int count = 0;
            for (int c : C) {
                if (c == label) {
                    count++;
                }
            }
            pc[i] = (double) count / C.length;
        }

        return new Matrix(pc);
    }
}

File: jar/src/main/kotlin/jline/api/trace/Mtrace_pc.kt
package jline.api.trace

import jline.util.matrix.Matrix

/**
 * Computes the probabilities of arrival for each class.
 * 
 * @param T the inter-arrival times (ignored, for orthogonality with other APIs)
 * @param C the class labels
 * @return column vector of probabilities for each class
 */
fun mtrace_pc(T: DoubleArray, C: IntArray): Matrix {
    val labels = C.distinct().sorted()
    val m = labels.size
    
    val pc = DoubleArray(m)
    
    for (i in 0 until m) {
        pc[i] = C.count { it == labels[i] }.toDouble() / C.size
    }
    
    return Matrix(pc)
}
/**
 * Mtrace Pc algorithms
 */
@Suppress("unused")
class MtracePcAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}